src/libraries/System.Text.Json/src/System/Text/Json/Document/JsonDocument.MetadataDb.cs

```diff
@@ -165,23 +146,52 @@ public void Dispose()
                 Length = 0;
             }
 
-            internal void TrimExcess()
+            /// <summary>
+            /// If using array pools, trim excess if necessary.
+            /// If not using array pools, release the temporary array pool and alloc.
```

Comment:
Is it possible/worth it to avoid the temporary ArrayPool rent to begin with?

Reply:
Since it can grow based on length, using the temp pool has advantages since it prevents allocs in those cases. When creating the "real" (non-pooled) buffer, it is created with the correct (trimmed) size which also prevents an alloc if we wanted to trim the size.

However, it is possible to add a mode that we try to detect whether small JSON will fit into a default buffer size, and not trim it afterwards. For example, if we know we are deserializing a number or string and not an object or array.

Reply:
I update the PR to address this as well as other improvements including caching of literal values.